File: src/backend/app/api/videos.py
from __future__ import annotations

from pathlib import Path

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from fastapi.responses import FileResponse
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.schemas.video_schema import VideoOut
from app.services import video_service

router = APIRouter(prefix="/api/videos", tags=["videos"])


@router.post("/upload", response_model=VideoOut)
def upload_video(file: UploadFile = File(...), db: Session = Depends(get_db)):
    if not file.filename:
        raise HTTPException(status_code=400, detail="Missing filename.")
    return video_service.save_upload(db, file)


@router.get("", response_model=list[VideoOut])
def list_videos(db: Session = Depends(get_db)):
    return video_service.list_videos(db)


@router.get("/{video_id}", response_model=VideoOut)
def get_video(video_id: str, db: Session = Depends(get_db)):
    return video_service.get_video(db, video_id)


@router.delete("/{video_id}")
def delete_video(video_id: str, db: Session = Depends(get_db)):
    video_service.delete_video(db, video_id)
    return {"deleted": True}


@router.get("/{video_id}/thumbnail")
def get_thumbnail(video_id: str, db: Session = Depends(get_db)):
    v = video_service.get_video(db, video_id)
    if not v.thumbnail_path:
        raise HTTPException(status_code=404, detail="Thumbnail not available.")
    p = Path(v.thumbnail_path)
    if not p.exists():
        raise HTTPException(status_code=404, detail="Thumbnail file missing on disk.")
    return FileResponse(str(p), media_type="image/jpeg")

File: src/frontend/src/api/videos.ts
import { api } from "./client";
import type { Video } from "../types/video";

export async function uploadVideo(file: File): Promise<Video> {
  const form = new FormData();
  form.append("file", file);
  const res = await api.post<Video>("/api/videos/upload", form, {
    headers: { "Content-Type": "multipart/form-data" },
  });
  return res.data;
}

export async function listVideos(): Promise<Video[]> {
  const res = await api.get<Video[]>("/api/videos");
  return res.data;
}

export async function getVideo(videoId: string): Promise<Video> {
  const res = await api.get<Video>(`/api/videos/${videoId}`);
  return res.data;
}

export async function deleteVideo(videoId: string): Promise<void> {
  await api.delete(`/api/videos/${videoId}`);
}

export function thumbnailUrl(videoId: string): string {
  return `${api.defaults.baseURL}/api/videos/${videoId}/thumbnail`;
}

